### ISSUE TITLE
Multithreading causes Travis builds to stall randomly

### ISSUE THREAD

**Issue by Bilge** [infection/infection#827]
| Question    | Answer
| ------------| ---------------
| Infection version | unknown
| Test Framework version | PHPUnit 7.5.17
| PHP version | 7.2.15
| Platform    | Ubuntu
| Github Repo | https://github.com/Porter-connectors/HttpConnector

This is an intermittent problem that does not affect every build nor does it happen every time. It typically affects around 10% of builds but cannot be reproduced reliably. For an example, see [build](https://travis-ci.org/Porter-connectors/HttpConnector/jobs/611272462). Infection appears to complete normally but the process stalls instead of exiting, causing the build to fail.

The relevant part of the infection invocation is `--threads=$(nproc)`.

>Metrics:
>         Mutation Score Indicator (MSI): 70%
>         Mutation Code Coverage: 76%
>         Covered Code MSI: 91%

>Please note that some mutants will inevitably be harmless (i.e. false positives).

>Dashboard report has not been sent: expected branch "master", found "5.0-rc"

>Time: 29s. Memory: 14.00MB

>No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
>Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

>The build has been terminated

**Comment by sanmai** [comment]
It is actually Infection that hangs.

**Comment by Bilge** [comment]
[Build](https://travis-ci.org/Porter-connectors/HttpConnector/jobs/611702313):
>The command "timeout -k 60s 40s vendor/bin/infection --min-msi=69 --threads=$(nproc) --coverage=build/coverage" exited with 124.

`man timeout`
> If  the  command times out, and --preserve-status is not set, then exit with status 124.

**Comment by Bilge** [comment]
However, I have reproduced this locally even without `-j`/`--threads`, so it might be due to something else than multithreading.

**Comment by sanmai** [comment]
Yeah, exit status 124 is a failure, so is the timeout. It should not behave like so.

**Comment by sanmai** [comment]
If you code (the code you're testing) _forks_ at some point, this could be a potential issue. Haven't had the time to look if it does anything like that.

**Comment by Bilge** [comment]
It spawns a web server (`php -S`) and a [stunnel](https://www.stunnel.org/) with [Symfony/Process](https://github.com/symfony/process).

**Comment by sanmai** [comment]
Sure, because Infection changes your code in many different ways, and some of the changes lead to this issue. I don't think there's anything we can do Infection-wise, but if you can exclude/hide these integration tests from Infection (say, with PHPUnit groups), it should certainly help. Mutation testing and integration tests don't play along very well, in principle.

I think we can let the issue linger for a time, but overall this is a "won't fix".

**Comment by theofidry** [comment]
@Bilge it is quite tricky whenever you are running Infection on code which depends on I/O, DB, network or other. Since by design Infection introduces a bug, whenever your code has outside interactions well you can no longer guarantee those interactions are done safely.

A maybe extreme and unrealistic scenario would be if you have a code removing a specific path, imagine infection replacing that path with `/` (root)... It could end up badly. Same goes for a DB: maybe infection introduces a bug that badly corrupts the DB or prevents a transaction rollbac...

TL:DR; whenever there is side-effect, you need an extra level of cautiousness. Sometimes just running it in a docker container can be enough, sometimes it's just better to exclude that code from Infection
